File: Prog_lab7_cpp/teacher.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#include <string.h>
#include <malloc.h>
#include "teacher.h"

//Метод установки рабочего стажа
bool teacher::SetWorkExp(int buf)
{
	if (buf < 0 || buf > 60)
		return (true);
	else
	{
		WorkExp = buf;
		return (false);
	}
}

//Метод установки ученой степени
bool teacher::SetDegree(char buf_string[])
{
	//обрезка входной строки
	if (strlen(buf_string) > 40)
		buf_string[40 - 1] = '\0';

	//проверка на ввод пустой строки
	if (strlen(buf_string) == 0)
		return (true);

	//определение массива недопустимых символов
	char invalid_symbols[] = "!@#$%^&*()_+1234567890-=\"№;:?*,./'][{}<>~`";
	int invalid_symbols_lenght = 42;

	//посимвольное сравнение входной строки и массива недопустимых символов
	int buf_len = strlen(buf_string);
	for (int i = 0; i < buf_len; i++)
		for (int j = 0; j < invalid_symbols_lenght; j++)
			if (buf_string[i] == invalid_symbols[j])
				return(true);

	//копирование входной строки в поле объекта,
	//при успешном прохождении проверки
	strcpy(Degree, buf_string);
	return (false);
}

//Метод установки названия факультета
bool teacher::SetFacultyName(char buf_string[])
{
	//обрезка входной строки
	if (strlen(buf_string) > 30)
		buf_string[30 - 1] = '\0';

	//проверка на ввод пустой строки
	if (strlen(buf_string) == 0)
		return (true);

	//определение массива недопустимых символов
	char invalid_symbols[] = "!@#$%^&*()_+1234567890-=\"№;:?*,./'][{}<>~` ";
	int invalid_symbols_lenght = 43;

	//посимвольное сравнение входной строки и массива недопустимых символов
	int buf_len = strlen(buf_string);
	for (int i = 0; i < buf_len; i++)
		for (int j = 0; j < invalid_symbols_lenght; j++)
			if (buf_string[i] == invalid_symbols[j])
				return(true);

	//копирование входной строки в поле объекта,
	//при успешном прохождении проверки
	strcpy(FacultyName, buf_string);
	return (false);
}

//Метод извлечения рабочего стажа
int teacher::GetWorkExp()
{
	int buf = WorkExp;
	return (buf);
}

//Метод извлечения ученой степени
char* teacher::GetDegree()
{
	char* buf_string = (char*)calloc(40, sizeof(char));
	strcpy(buf_string, Degree);
	return (buf_string);
}

//Метод извлечения названия факультета
char* teacher::GetFacultyName()
{
	char* buf_string = (char*)calloc(30, sizeof(char));
	strcpy(buf_string, FacultyName);
	return (buf_string);
}

//Метод  инициализации объекта класса
bool teacher::Init(int buf_WorkExp, char buf_Degree[], char buf_FacultyName[], human buf_human)
{
	//создаём объект класса human для проверки формата входных данных
	teacher check;

	//проверка на формат ввода human??????????
	if (check.SetWorkExp(buf_WorkExp) || check.SetDegree(buf_Degree) || check.SetFacultyName(buf_FacultyName))
		return (true);
	else
	{
		SetWorkExp(buf_WorkExp);
		SetDegree(buf_Degree);
		SetFacultyName(buf_FacultyName);
		SetFacultyName(check.GetFacultyName());
		thehuman = buf_human;
		return (false);
	}
}

//Метод ввода полей объекта класса из консоли
bool teacher::Read()
{
	//создаём объект класса teacher для проверки формата входных данных
	teacher check;

	//буферный переменная типа int 
	//для проверки формата входных данных 
	int buf_int;

	//проверка на формат входных данных
	printf("Enter working experience:\n");
	if (scanf("%d", &buf_int) != 1)
		return (true);
	else
	{
		//попытка записи в поле объекта
		if (check.SetWorkExp(buf_int))
			return (true);
		else
		{
			//создаём буферную строку для проверки формата входных данных
			char buf_string[40];
			rewind(stdin);
			printf("Enter scientific degree:\n");
			fgets(buf_string, 40, stdin);
			//если строка короче 40, то заменить "перенос на новую строку" "концом строки"
			if (buf_string[strlen(buf_string) - 1] == '\n')
				buf_string[strlen(buf_string) - 1] = '\0';
			//если строка длиннее 40, то очистить входной поток
			else rewind(stdin);

			//попытка записи в поле объекта
			if (check.SetDegree(buf_string))
				return (true);
			else
			{
				printf("Enter faculty name:\n");
				fgets(buf_string, 30, stdin);
				//если строка короче 30, то заменить "перенос на новую строку" "концом строки"
				if (buf_string[strlen(buf_string) - 1] == '\n')
					buf_string[strlen(buf_string) - 1] = '\0';
				//если строка длиннее 30, то очистить входной поток
				else rewind(stdin);

				//попытка записи в поле объекта
				if (check.SetFacultyName(buf_string))
					return (true);
				else
				{
					//попытка записи в объект
					if (check.thehuman.Read())
						return (true);

					//когда все проверки пройдены,
					//можно перенести данные в главный объект
					else
					{
						SetWorkExp(check.GetWorkExp());
						SetDegree(check.GetDegree());
						SetFacultyName(check.GetFacultyName());
						thehuman = check.thehuman;
						return(false);
					}
				}
			}
		}
	}
}

//Метод вывода содержимого объекта класса в консоль
void teacher::Display()
{
	printf("working experience: %d years\n", WorkExp);
	printf("scientific degree: %s\n",Degree);
	printf("faculty name: %s\n", FacultyName);
	thehuman.Display();
}

File: Prog_lab7_java/Teacher.java
package Prog_lab5;

import java.util.*;

public class Teacher 
{
	//Attributes
	private int workExp;
	private String degree;
	private String facultyName;
	public Human humanField = new Human();

	//Methods
	public boolean setWorkExp(int buf)
	{
		if (buf < 0 || buf > 60)
		return (true);
		else
		{
			workExp = buf;
			return (false);
		}
	}

	public boolean setDegree(String bufString)
	{
		if (bufString.isEmpty()) 
			return (true);
		
		String invalidSymbStr = "!@#$%^&*()_+1234567890-=\";:?*,./'][{}<>~`";
		char[] invalidSymbols = invalidSymbStr.toCharArray();
		for (char symb : invalidSymbols) 
		{
			if(bufString.indexOf(symb) != (-1))
				return (true);
		}

		degree = degree.copyValueOf(bufString.toCharArray());
		return (false);
	}

	public boolean setFacultyName(String bufString)
	{
		if (bufString.isEmpty()) 
			return (true);
		
		String invalidSymbStr = "!@#$%^&*()_+1234567890-=\";:?*,./'][{}<>~` ";
		char[] invalidSymbols = invalidSymbStr.toCharArray();
		for (char symb : invalidSymbols) 
		{
			if(bufString.indexOf(symb) != (-1))
				return (true);
		}

		facultyName = facultyName.copyValueOf(bufString.toCharArray());
		return (false);
	}

	public int getWorkExp()
	{
		return (workExp);
	}

	public String getDegree()
	{
		return (String.copyValueOf(degree.toCharArray()));
	}

	public String getFacultyName()
	{
		return (String.copyValueOf(facultyName.toCharArray()));
	}
	
	public boolean init(int bufWorkExp, String bufDegree, String bufFacultyName, Human bufHuman)
	{
		Teacher check = new Teacher();

		if (check.setWorkExp(bufWorkExp) || check.setDegree(bufDegree) || check.setFacultyName(bufFacultyName))
			return (true);
		else
		{
			this.setWorkExp(bufWorkExp);
			this.setDegree(bufDegree);
			this.setFacultyName(bufFacultyName);
			this.humanField.init(bufHuman.getId(), bufHuman.getAge(),
			bufHuman.getHeight(), bufHuman.getWeight(),bufHuman.getGender(), bufHuman.fioField);
			return (false);
		}
	}
	
	public boolean read()
	{
		Teacher check = new Teacher();
		Scanner scanner = new Scanner(System.in);

		System.out.print("Enter scientific degree:\n");
		if (!scanner.hasNextLine())
			return (true);
		if (check.setDegree(scanner.nextLine()))
			return (true);


		System.out.print("Enter faculty name:\n");
		if (!scanner.hasNextLine())
			return (true);
		if (check.setFacultyName(scanner.nextLine()))
			return (true);


		System.out.print("Enter working experience:\n");
		if (!scanner.hasNextInt())
			return (true);
		if (check.setWorkExp(scanner.nextInt()))
			return (true);


		if (check.humanField.read())
			return(true);


		this.init(check.getWorkExp(), check.getDegree(),
				check.getFacultyName(), check.humanField);
		return(false);
	}

	public void display()
	{
		System.out.printf("working experience: %d years\n", workExp);
		System.out.printf("scientific degree: %s\n",degree);
		System.out.printf("faculty name: %s\n", facultyName);
		humanField.display();
	}
}